Add configuration for and handling of other image types

FinnReinhardtBsc opened the pull request:
* Add multiselect to enable / disable specific image types
* Adjust config to account for different image types
* Add additional config for new image types (and filter image ids accordingly)
* Adjust logic to account for new image types

Based on changes from [!2](https://github.com/basecom/magento2-live-search-image-urls/pull/2)

FinnReinhardtBsc committed:
Add SelectedImageTypes.php to enable / disable resizing

FinnReinhardtBsc committed:
Add virtual types to filter images by type

FinnReinhardtBsc committed:
Update system.xml with missing configurations

FinnReinhardtBsc committed:
Update ModuleConfig.php to retrieve new configurations

FinnReinhardtBsc committed:
Update config.xml with additional default values

FinnReinhardtBsc committed:
Add image type specific fallback values for width and height

FinnReinhardtBsc committed:
Update SetCachedImageUrls.php logic to take new configuration into account

FinnReinhardtBsc committed:
Add filter entries without specified width / height

FinnReinhardtBsc committed:
Reference constant instead of hardcoding 'none'

FinnReinhardtBsc committed:
Rename constant

FinnReinhardtBsc committed:
Update CHANGELOG.md

FinnReinhardtBsc committed:
Update composer.json version

FinnReinhardtBsc committed:
Update README.md

pykettk commented:
@FinnReinhardtBsc Looks good to me. I think the only things I'd say aren't massively important:
1. It'd be nice to type hint the `$imageType` arguments in `System/ModuleConfig.php`
2. There's probably no need to use `CDATA` in the comment of the `image_types` field in `system.xml`
3. Some of the `system.xml` fields have had `canRestore` set to `0` despite there being entries for those fields in `config.xml`. Providing sensible defaults and allowing merchants to revert to using them is a nice quality of life feature IMO.

FinnReinhardtBsc committed:
Add string as explicit type to parameter in ModuleConfig.php

FinnReinhardtBsc commented:
Thanks for your review, @pykettk! 

> 1. It'd be nice to type hint the `$imageType` arguments in `System/ModuleConfig.php`

Good eye, I added the missing types there! 👍 

> 2. There's probably no need to use `CDATA` in the comment of the `image_types` field in `system.xml`

I removed CDATA for the comment field.

> 3. Some of the `system.xml` fields have had `canRestore` set to `0` despite there being entries for those fields in `config.xml`. Providing sensible defaults and allowing merchants to revert to using them is a nice quality of life feature IMO.

I see your point. When configuring this extension, I personally found it rather annoying to always uncheck the "use system value" box. As I see it, the default for the selected image types and resize mode provide less value than the other defaults do, because it depends a lot on your system.

But I understand that it is a good quality of life feature to have, so I changed them again :)

---

I also removed the "Only works on Hyvä Storefronts!" text (as mentioned [here](https://github.com/basecom/magento2-live-search-image-urls/pull/2#discussion_r2011766983)). My testing showed, that this functionality is not limited to Hyvä-Themes (although I would always recommend people to use Hyvä, of course :D). I just wanted to hear your take on this, in case I missed something.

Thanks for your review, @pykettk!
If there aren't any other issues, I will merge the changes and create a new release 🎉

FinnReinhardtBsc committed:
Remove CDATA from comments that do not require it

pykettk commented:
> I also removed the "Only works on Hyvä Storefronts!" text

I think that was my bad - I copied part of one of our `system.xml` files when working on this and overlooked it 😅 

Everything else looks good though! 💪

FinnReinhardtBsc committed:
Update CHANGELOG.md and composer.json version number
